File: app/src/main/java/com/example/recipeclipper/data/local/dao/GroceryDao.kt
package com.example.recipeclipper.data.local.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Transaction
import com.example.recipeclipper.data.local.entity.GroceryItemEntity
import kotlinx.coroutines.flow.Flow

/**
 * A planned recipe's ingredients, for "Add this week's ingredients" (#50): the entry's day and
 * servings with the recipe's lines, yield and language.
 */
data class PlannedIngredientsRow(
    val entryId: Long,
    val day: Long,
    val servings: Int?,
    val recipeId: Long,
    val title: String,
    val ingredients: List<String>,
    val yield: String?,
    val language: String?
)

/** The grocery list (#50). One list for now ([GroceryItemEntity.DEFAULT_LIST]). */
@Dao
abstract class GroceryDao {

    @Query("SELECT * FROM grocery_items WHERE listId = :listId ORDER BY sortOrder ASC, id ASC")
    abstract fun observeItems(listId: Long = GroceryItemEntity.DEFAULT_LIST): Flow<List<GroceryItemEntity>>

    @Query("SELECT * FROM grocery_items WHERE id IN (:ids)")
    abstract suspend fun items(ids: List<Long>): List<GroceryItemEntity>

    @Query("SELECT * FROM grocery_items WHERE listId = :listId AND checked = 1")
    abstract suspend fun checkedItems(listId: Long = GroceryItemEntity.DEFAULT_LIST): List<GroceryItemEntity>

    @Query("SELECT COALESCE(MAX(sortOrder), -1) + 1 FROM grocery_items WHERE listId = :listId")
    protected abstract suspend fun nextOrder(listId: Long): Int

    @Insert
    protected abstract suspend fun insert(items: List<GroceryItemEntity>)

    /** Adds [items] at the end of their list, in the order given, whatever their sortOrder said. */
    @Transaction
    open suspend fun add(items: List<GroceryItemEntity>) {
        if (items.isEmpty()) return
        val start = nextOrder(items.first().listId)
        insert(items.mapIndexed { index, item -> item.copy(id = 0, sortOrder = start + index) })
    }

    /** Undoes a delete: the same rows back, ids, uids and places included. */
    @Insert
    abstract suspend fun restore(items: List<GroceryItemEntity>)

    @Query("UPDATE grocery_items SET checked = :checked, updatedAt = :now WHERE id IN (:ids)")
    abstract suspend fun setChecked(ids: List<Long>, checked: Boolean, now: Long)

    @Query("UPDATE grocery_items SET aisle = :aisle, updatedAt = :now WHERE id IN (:ids)")
    abstract suspend fun setAisle(ids: List<Long>, aisle: String, now: Long)

    @Query("DELETE FROM grocery_items WHERE id IN (:ids)")
    abstract suspend fun delete(ids: List<Long>)

    /**
     * Every recipe planned from day [start] to [end] inclusive, in plan order (as the Week shows
     * it). Notes have no ingredients, so they are left out.
     */
    @Query(
        """
        SELECT e.id AS entryId, e.day, e.servings, r.id AS recipeId, r.title, r.ingredients,
               r.servings AS yield, r.language
        FROM meal_plan_entries e
        JOIN meal_types t ON t.id = e.mealTypeId
        JOIN recipes r ON r.id = e.recipeId
        WHERE e.day BETWEEN :start AND :end
        ORDER BY e.day ASC, t.sortOrder ASC, t.id ASC, e.sortOrder ASC, e.id ASC
        """
    )
    abstract suspend fun plannedIngredients(start: Long, end: Long): List<PlannedIngredientsRow>
}

File: ios/RecipeClipper/Data/Local/GroceryDao.swift
import Foundation

/// The grocery list's SQL (#50), ported from Android's GroceryDao. Synchronous; run inside
/// `AppDatabase.read` / `write` (a write is one transaction).
struct GroceryDao {
    let db: SQLiteConnection

    func items(listId: Int64 = GroceryItemRecord.defaultList) throws -> [GroceryItemRecord] {
        try db.query(
            "SELECT \(GroceryItemRecord.columns) FROM grocery_items WHERE listId = ? ORDER BY sortOrder ASC, id ASC",
            listId, map: GroceryItemRecord.init(row:)
        )
    }

    func items(ids: [Int64]) throws -> [GroceryItemRecord] {
        try ids.compactMap { id in
            try db.queryOne(
                "SELECT \(GroceryItemRecord.columns) FROM grocery_items WHERE id = ?", id,
                map: GroceryItemRecord.init(row:)
            )
        }
    }

    func checkedItems(listId: Int64 = GroceryItemRecord.defaultList) throws -> [GroceryItemRecord] {
        try db.query(
            "SELECT \(GroceryItemRecord.columns) FROM grocery_items WHERE listId = ? AND checked = 1 ORDER BY sortOrder ASC, id ASC",
            listId, map: GroceryItemRecord.init(row:)
        )
    }

    /// Adds `items` at the end of their list, in the order given, whatever their sortOrder said.
    /// Call inside a write.
    func add(_ items: [GroceryItemRecord]) throws {
        guard let first = items.first else { return }
        let start = try db.queryOne(
            "SELECT COALESCE(MAX(sortOrder), -1) + 1 FROM grocery_items WHERE listId = ?", first.listId
        ) { $0.int(0) } ?? 0
        for (index, item) in items.enumerated() {
            var row = item
            row.id = 0
            row.sortOrder = start + index
            try insert(row)
        }
    }

    /// An `id` of 0 lets SQLite assign one; a non-zero id is honoured (what undo relies on).
    /// A recipe that went meanwhile leaves the item without a source rather than failing.
    func insert(_ e: GroceryItemRecord) throws {
        try db.run(
            """
            INSERT INTO grocery_items (\(GroceryItemRecord.columns))
            VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7,
                    CASE WHEN EXISTS (SELECT 1 FROM recipes WHERE id = ?8) THEN ?8 END, ?9, ?10, ?11)
            """,
            e.id == 0 ? nil : e.id, e.listId, e.text, e.language, e.aisle, e.checked, e.sortOrder,
            e.recipeId, e.plannedDay, e.updatedAt, e.uid
        )
    }

    /// Undoes a delete: the same rows back, ids, uids and places included.
    func restore(_ items: [GroceryItemRecord]) throws {
        for item in items { try insert(item) }
    }

    func setChecked(_ ids: [Int64], checked: Bool, now: Int64) throws {
        for id in ids {
            try db.run("UPDATE grocery_items SET checked = ?, updatedAt = ? WHERE id = ?", checked, now, id)
        }
    }

    func setAisle(_ ids: [Int64], aisle: String, now: Int64) throws {
        for id in ids {
            try db.run("UPDATE grocery_items SET aisle = ?, updatedAt = ? WHERE id = ?", aisle, now, id)
        }
    }

    func delete(_ ids: [Int64]) throws {
        for id in ids { try db.run("DELETE FROM grocery_items WHERE id = ?", id) }
    }

    /// Every recipe planned from day `start` to `end` inclusive, in plan order (as the Week
    /// shows it). Notes have no ingredients, so they are left out.
    func plannedIngredients(start: Int64, end: Int64) throws -> [PlannedIngredients] {
        try db.query(
            """
            SELECT e.id, e.day, e.servings, r.id, r.title, r.ingredients, r.servings, r.language
            FROM meal_plan_entries e
            JOIN meal_types t ON t.id = e.mealTypeId
            JOIN recipes r ON r.id = e.recipeId
            WHERE e.day BETWEEN ? AND ?
            ORDER BY e.day ASC, t.sortOrder ASC, t.id ASC, e.sortOrder ASC, e.id ASC
            """,
            start, end
        ) { row in
            PlannedIngredients(
                entryId: row.int64(0),
                day: row.int64(1),
                servings: row.isNull(2) ? nil : row.int(2),
                recipeId: row.int64(3),
                title: row.string(4),
                ingredients: JSONColumns.decodeStrings(row.string(5)),
                yield: row.optionalString(6),
                language: row.optionalString(7)
            )
        }
    }
}
